stonesoup/updater/kalman.py

```diff
@@ -628,3 +653,135 @@ def update(self, hypothesis, **kwargs):
+    Control Systems, Vol. 3, 1966, pp. 293–340
+
+    [2] Zanetti, R. & D’Souza, C. (2013). Recursive Implementations of the Schmidt-Kalman
+    ‘Consider’   Filter. The Journal of the Astronautical Sciences. 60. 672-685.
+    10.1007/s40295-015-0068-7.
+
+    """
+    consider: np.ndarray = Property(default=None,
+                                    doc="The boolean vector of 'consider' parameters. True "
+                                        "indicates considered, False are state parameters to be "
+                                        "estimated. If undefined these default to all False, i.e."
+                                        "the standard Kalman filter.")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.consider is None:
+            self.consider = np.zeros(self.measurement_model.ndim_state, dtype=bool)
+
+    def _posterior_mean(self, predicted_state, kalman_gain, measurement, measurement_prediction):
+        """Compute the posterior mean, :math:`s_{k|k} = s_{k|k-1} + K_s (z - H_s s_{k|k-1} -
+        H_p p_{k|k-1})`, :math:`p_{k|k} = p_{k|k-1}.
+
+        Parameters
+        ----------
+        predicted_state : :class:`State`, :class:`Prediction`
+            The predicted state
+        kalman_gain : numpy.ndarray
+            The reduced form of the Kalman gain, :math:`K_s`
+         measurement : :class:`Detection`
+            The measurement
+        measurement_prediction : :class:`MeasurementPrediction`
+            Predicted measurement
+
+        Returns
+        -------
+        : :class:`StateVector`
+            The posterior mean estimate
+        """
+        post_mean = predicted_state.state_vector.copy()
+        post_mean[np.ix_(~self.consider)] += \
+            kalman_gain @ (measurement.state_vector - measurement_prediction.state_vector)
+        return post_mean.view(StateVector)
+
+    def _posterior_covariance(self, hypothesis):
+        """
+        Return the posterior covariance for a given hypothesis
+
+        Parameters
+        ----------
+        hypothesis: :class:`~.Hypothesis`
+            A hypothesised association between state prediction and measurement. It returns the
+            measurement prediction which in turn contains the measurement cross covariance,
+            :math:`P_{k|k-1} H_k^T and the innovation covariance,
+            :math:`S = H_k P_{k|k-1} H_k^T + R`
+
+        Returns
+        -------
+        : :class:`~.CovarianceMatrix`
+            The posterior covariance matrix rendered via the Schmidt-Kalman update process.
+        : numpy.ndarray
+            The reduced form of the Kalman gain,
+            :math:`K_s = (P_{ss,k|k-1} H_{s,k}^T + P_{sp,k|k-1} H_{p,k}^T) S^{-1}`
+
+        """
+        # Intermediate matrices P_p and H.
+        pp = hypothesis.prediction.covar[np.ix_(range(0, len(self.consider)), self.consider)]
```

Comment:
Cool - how do you get line-by-line profiling? I only managed to resolve at the method level.

Reply:
In a not very elegant way of just commenting out the lines and seeing the run time. Just to show difference, I tried modifying the code to pass in pre-calculated `pp` value.
![image](https://github.com/dstl/Stone-Soup/assets/19266036/51cb7dc1-1884-405a-99b5-e4073b00cb32)

Reply:
Saves about 100ms in example above. Not sure what else to suggest.

```suggestion
        pp = hypothesis.prediction.covar[np.tile(self.consider, (len(self.consider), 1))]
        pp = pp.reshape((len(self.consider), np.sum(self.consider)))
```

Reply:
So I went with `pp = hypothesis.prediction.covar[:len(self.consider), self.consider]` because I'm told creating a view rather than a copy is more efficient. But that makes it slower?!

Reply:
No discernible difference with that new method.

Reply:
> So I went with `pp = hypothesis.prediction.covar[:len(self.consider), self.consider]` because I'm told creating a view rather than a copy is more efficient. But that makes it slower?!

That's because unless it's contiguous in the array, this is still a copy. So if you enforce the estimate parameters come first and then do `pp = hypothesis.prediction.covar[:, np.sum(~self.consider):]` then you can save (~30%) time by creating a view. 

But it feels like this is overly restrictive -- the user should be able to select which parameters are estimated and which considered. Unless someone wants to go down a cython route, it's probably time to move on..